src/crypto-api/key-passphrase.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2024 The Matrix.org Foundation C.I.C.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { randomString } from "../randomstring.ts";
+
+const DEFAULT_ITERATIONS = 500000;
+
+const DEFAULT_BITSIZE = 256;
+
+/* eslint-disable camelcase */
+interface IAuthData {
+    private_key_salt?: string;
+    private_key_iterations?: number;
+    private_key_bits?: number;
+}
+/* eslint-enable camelcase */
+
+interface IKey {
+    key: Uint8Array;
+    salt: string;
+    iterations: number;
+}
+
+/**
+ * Derive a key from a passphrase using the salt and iterations from the auth data.
+ * @param authData
+ * @param password
+ */
+export function keyFromAuthData(authData: IAuthData, password: string): Promise<Uint8Array> {
+    if (!authData.private_key_salt || !authData.private_key_iterations) {
+        throw new Error("Salt and/or iterations not found: " + "this backup cannot be restored with a passphrase");
+    }
+
+    return deriveKey(
+        password,
+        authData.private_key_salt,
+        authData.private_key_iterations,
+        authData.private_key_bits || DEFAULT_BITSIZE,
+    );
+}
+
+/**
+ * Derive a key from a passphrase.
+ * @param password
+ */
+export async function keyFromPassphrase(password: string): Promise<IKey> {
+    const salt = randomString(32);
+
+    const key = await deriveKey(password, salt, DEFAULT_ITERATIONS, DEFAULT_BITSIZE);
+
+    return { key, salt, iterations: DEFAULT_ITERATIONS };
+}
```

Comment:
This isn't something that applications should be calling directly, either (instead they should call `CryptoApi.createRecoveryKeyFromPassphrase`), so again let's not add it here.

Currently, it's only called from (a) bits of legacy crypto, and (b) `RustCrypto.createRecoveryKeyFromPassphrase`.

Since it's only two lines, I'd be inclined to inline it in `RustCrypto.createRecoveryKeyFromPassphrase`, and leave the function in its old place in `crypto/key_passphrase.ts` for use by the legacy crypto stack.

Reply:
btw, don't be afraid to create a second definition  of `DEFAULT_ITERATIONS` (probably with a better name) in rust-crypto. I don't think we need a second definition of `DEFAULT_BITSIZE`  -- we can just omit the parameter.